grakn-kb/src/main/java/ai/grakn/kb/internal/structure/Casting.java

```diff
@@ -43,21 +45,42 @@
  */
 public class Casting {
     private final EdgeElement edgeElement;
-    private final Cache<Role> cachedRoleType = new Cache<>(Cacheable.concept(), () -> (Role) edge().tx().getSchemaConcept(LabelId.of(edge().property(Schema.EdgeProperty.ROLE_LABEL_ID))));
-    private final Cache<RelationshipType> cachedRelationType = new Cache<>(Cacheable.concept(), () -> (RelationshipType) edge().tx().getSchemaConcept(LabelId.of(edge().property(Schema.EdgeProperty.RELATIONSHIP_TYPE_LABEL_ID))));
-
+    private final Cache<Role> cachedRole = new Cache<>(Cacheable.concept(), () -> (Role) edge().tx().getSchemaConcept(LabelId.of(edge().property(Schema.EdgeProperty.ROLE_LABEL_ID))));
     private final Cache<Thing> cachedInstance = new Cache<>(Cacheable.concept(), () -> edge().target().
             flatMap(vertexElement -> edge().tx().factory().<Thing>buildConcept(vertexElement)).
             orElseThrow(() -> GraknTxOperationException.missingRolePlayer(edge().id().getValue()))
     );
 
-    private final Cache<Relationship> cachedRelation = new Cache<>(Cacheable.concept(), () -> edge().source().
+    private final Cache<Relationship> cachedRelationship = new Cache<>(Cacheable.concept(), () -> edge().source().
             flatMap(vertexElement -> edge().tx().factory().<Relationship>buildConcept(vertexElement)).
             orElseThrow(() -> GraknTxOperationException.missingRelationship(edge().id().getValue()))
     );
 
-    public Casting(EdgeElement edgeElement){
+    private final Cache<RelationshipType> cachedRelationshipType = new Cache<>(Cacheable.concept(), () -> {
+        if(cachedRelationship.isPresent()){
+            return cachedRelationship.get().type();
+        } else {
+            return (RelationshipType) edge().tx().getSchemaConcept(LabelId.of(edge().property(Schema.EdgeProperty.RELATIONSHIP_TYPE_LABEL_ID)));
+        }
+    });
+
+    private Casting(EdgeElement edgeElement, @Nullable Relationship relationship, @Nullable Role role, @Nullable Thing thing){
         this.edgeElement = edgeElement;
+        if(relationship != null) this.cachedRelationship.set(relationship);
+        if(role != null) this.cachedRole.set(role);
+        if(thing != null) this.cachedInstance.set(thing);
+    }
```

Comment:
Alternative suggestion: given these caches are mutable anyway, maybe expose some setters for them? Then you only need one constructor and a setter for each cache. I'm not 100% sold on either way though.

Reply:
Nothing else sets these caches explicitly that's why I don't think it's worth making setters.